File: 37.解数独.c
/*
 * @lc app=leetcode.cn id=37 lang=c
 *
 * [37] 解数独
 */

// @lc code=start
bool isVaild(char **board, int row, int col, char str, int boardSize, int *boardColSize)
{

  int i, j;
  for (i = 0; i < boardColSize[0]; i++)
  {
    if (board[row][i] == str)
      return false;
  }

  for (i = 0; i < boardSize; i++)
  {
    if (board[i][col] == str)
      return false;
  }

  for (i = (row / 3) * 3; i < (row / 3) * 3 + 3; i++)
  {
    for (j = (col / 3) * 3; j < (col / 3) * 3 + 3; j++)
    {
      if (board[i][j] == str)
        return false;
    }
  }

  return true;
}

bool back(char **board, int boardSize, int *boardColSize)
{
  int i, j, k;
  for (i = 0; i < boardColSize[0]; i++)
  {
    for (j = 0; j < boardSize; j++)
    {
      if (board[j][i] != '.')
        continue;
      for (k = 1; k <= 9; k++)
      {
        if (isVaild(board, j, i, k + '0', boardSize, boardColSize))
        {
          board[j][i] = k + '0';
          if (back(board, boardSize, boardColSize))
            return true;
          board[j][i] = '.';
        }
      }
      return false;
    }
  }
  return true;
}

void solveSudoku(char **board, int boardSize, int *boardColSize)
{
  back(board, boardSize, boardColSize);
}
// @lc code=end

File: store/37.解数独.js
/*
 * @lc app=leetcode.cn id=37 lang=javascript
 *
 * [37] 解数独
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */

let res;

const isVaild = (row, col, val, matrix) => {
  for (let i = 0; i < matrix.length; i++) {
    if (val === matrix[i][col]) return false;
  }

  for (let j = 0; j < matrix.length; j++)
    if (val === matrix[row][j]) return false;

  const startRow = Math.floor(row / 3) * 3;
  const startCol = Math.floor(col / 3) * 3;

  for (let i = startRow; i < startRow + 3; i++) {
    for (let j = startCol; j < startCol + 3; j++) {
      if (matrix[i][j] === val) return false;
    }
  }

  return true;
};

const trackback = (matrix) => {
  for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix[0].length; j++) {
      if (matrix[i][j] !== ".") continue;
      for (let k = 1; k <= 9; k++) {
        if (isVaild(i, j, `${k}`, matrix)) {
          matrix[i][j] = `${k}`;
          if (trackback(matrix)) return true;
          matrix[i][j] = ".";
        }
      }
      return false;
    }
  }
  return true;
};

var solveSudoku = function (board) {
  trackback(board);
};
// @lc code=end
